### DISCUSSION TITLE
vscode useExhaustiveDependencies errors does not have Quick Fix

### DISCUSSION THREAD

**Discussion by oddball** [biomejs/biome#1928] [Q&A]
I have been checking if biome can replace eslint and prettier for me.

Ran into 3 issues that for me are showstoppers:

 - [ ] 1. When having a useEffect with incorrect dependencies, the only QuickFix available in vscode is "Suppress rule lint/correctness/useExhaustiveDependencies", when what I want is an option to calculate what the dependencies should be, and set it. Is that not possible? It is unreasonable to hand code the dependency lists. I have many hundreds of useEffects
 - [x] 2. I have a number of "This hook specifies a dependency more specific that its captures", and I cannot turn that off without turning the "lint/correctness/useExhaustiveDependencies" off.
 - [x] 3. I got a number of "This hook does not specify all of its dependencies", when the useEffect, actually do that. Typically "value" is in the dependency list, but the hook uses "value.whatever". And again, I can not turn it off, without turning the "lint/correctness/useExhaustiveDependencies" off.

Any suggestions would be most appreciated!

**Comment by arendjr** [comment]
> what I want is an option to calculate what the dependencies should be, and set it. Is that not possible?

I don’t think we currently support that, no. We could implement it, but it would always be an unsafe fix only, because there are many situations why people sometimes *intentionally* deviate from the expected list of dependencies.

> It is unreasonable to hand code the dependency lists. I have many hundreds of useEffects

Just to verify, are those all your `useEffect()` usages, or only the ones that need fixing? How did you handle it previously with ESLint? Assuming the other issues get fixed, I would expect you only need to update the suppression comments, no?

> I have a number of "This hook specifies a dependency more specific that its captures", and I cannot turn that off without turning the "lint/correctness/useExhaustiveDependencies" off.

Do you have an example of this one?

> I got a number of "This hook does not specify all of its dependencies", when the useEffect, actually do that. Typically "value" is in the dependency list, but the hook uses "value.whatever". And again, I can not turn it off, without turning the "lint/correctness/useExhaustiveDependencies" off.

It sounds like a bug, but do you have an example of some code that triggers this? It might be best to create an issue for it.

**Reply by rfdrake27** [reply]
> > what I want is an option to calculate what the dependencies should be, and set it. Is that not possible?
> 
> I don’t think we currently support that, no. We could implement it, but it would always be an unsafe fix only, because there are many situations why people sometimes _intentionally_ deviate from the expected list of dependencies.
> 

Just chiming in to say that issue 1 of the original post was the singular reason why I turned off Biome linting and went back to ESLint. I understand but disagree with your reasoning here. In the spirit of linting, it makes sense that the linter would try to enforce the complete dependency list with the Quick Fix. In the vast majority of cases that is what _should_ be done. There are rare times you do have to omit a dependency, but adding the comment to disable the rule for that line is, in my opinion, the exact level of "are you sure there is no better way to do this?" reminder for bypassing the warning. The other 99% of the time, Biome's lack of the Quick Fix functionality for React deps is a major hit to productivity.

**Reply by ematipico** [reply]
I believe you both are saying the same thing in a different way.

We will gladly accept a PR to add this feature. Some help would be very welcome, as we are currently swamped on other fronts.

**Comment by oddball** [comment]
1 . It is implemented in ESLint( could possible be prettier). And for me, one of the features I cannot live without. In my case, I never calculate the dependency list myself. I would assume, that if I ever want to deviate from a complete list, then I would ignore just that line.

I got hundreds of dependency lists that needs fixing (according to biome). I dont think it is true, cause those are not hand written by me, they are calculated by ESLint(could possible be prettier).

2. I am not sure its a good example, but the first one I found. 

```typescript
  useEffect(() => {
    if (previousFormData && formData[dateName] !== previousFormData[dateName]) {
      const newQueryArgs = { ...queryArgs };
      newQueryArgs[dateName] = formData[dateName];
      navigate(location.pathname + "?" + queryString.stringify(newQueryArgs));
    } else if (queryArgs && dateName in queryArgs && formData[dateName] !== queryArgs[dateName]) {
      const change: any = {};
      change[dateName] = queryArgs[dateName];
      setFormData(change);
    }
  }, [dateName, previousFormData, formData, queryArgs, navigate, location]);
  ```
  
  ```bash
  ⚠ This hook specifies a dependency more specific that its captures: location

    35 │   const { queryArgs } = useQueryArgs();
    36 │
  > 37 │   useEffect(() => {
       │   ^^^^^^^^^
    38 │     if (previousFormData && formData[dateName] !== previousFormData[dateName]) {
    39 │       const newQueryArgs = { ...queryArgs };

  ℹ This capture is more generic than...

    40 │       newQueryArgs[dateName] = formData[dateName];
    41 │       navigate(location.pathname + "?" + queryString.stringify(newQueryArgs));
  > 42 │     } else if (queryArgs && dateName in queryArgs && formData[dateName] !== queryArgs[dateName]) {
       │                                                      ^^^^^^^^^^^^^^^^^^
    43 │       const change: any = {};
    44 │       change[dateName] = queryArgs[dateName];

  ℹ ...this dependency.

    45 │       setFormData(change);
    46 │     }
  > 47 │   }, [dateName, previousFormData, formData, queryArgs, navigate, location]);
       │                                                                  ^^^^^^^^
    48 │
    49 │   return (
```

3.  formData is in the list, therefor formData[dateName] should be covered.

```bash
  ⚠ This hook does not specify all of its dependencies: formData[dateName]

    35 │   const { queryArgs } = useQueryArgs();
    36 │
  > 37 │   useEffect(() => {
       │   ^^^^^^^^^
    38 │     if (previousFormData && formData[dateName] !== previousFormData[dateName]) {
    39 │       const newQueryArgs = { ...queryArgs };

  ℹ This dependency is not specified in the hook dependency list.

    40 │       newQueryArgs[dateName] = formData[dateName];
    41 │       navigate(location.pathname + "?" + queryString.stringify(newQueryArgs));
  > 42 │     } else if (queryArgs && dateName in queryArgs && formData[dateName] !== queryArgs[dateName]) {
       │                                                      ^^^^^^^^^^^^^^^^^^
    43 │       const change: any = {};
    44 │       change[dateName] = queryArgs[dateName];
```

**Reply by Sec-ant** [reply]
Re 2, I believe this issue has already been fixed, just hasn't published yet #1871, you can use the nightly channel, and a new stable release is on the way #1969.

**Reply by arendjr** [reply]
I also noticed there are actually some auto-fixes implemented too, though I'm not sure which version they were implemented... @oddball Now that 1.6 is out, could you mention which issues are still relevant?

**Reply by oddball** [reply]
@arendjr, I just checked with versions 1.6.1 and  number 2 and 3 seems gone. Thanks. Still no autofix of useEffect dependencies though.

**Reply by ttdatt** [reply]
I'm a big fan of this vscode quick fix It would be fantastic if you could add it as a feature. If you're concerned about potential issues, it could be opt-in setting that users can enable if they want

**Reply by oddball** [reply]
@ttdatt, I would prefer if the dependency list was automatically calculated unless the statement had a "// biome-ignore lint/correctness/useExhaustiveDependencies: need here or something"

**Reply by jordangarcia** [reply]
I agree that this isn't the most correct of fixes and it's kind of a blanket statement to say every variable in this should be a dependency of the `useEffect`.  However, I find this maybe the single must useful eslint rule and code action fix in practice.   Most of the time a useEffect lint rule is invalid the solution is to add all of the dependencies to the array.

for me and my company unless this code fix is implemented we'll probably be holding off on making the switch over from eslint :/

**Comment by dangus21** [comment]
any updates on this issue?

**Reply by ematipico** [reply]
Please read the threads: https://github.com/biomejs/biome/discussions/1928#discussioncomment-10582258

**Comment by maciejgrzybek-ramp** [comment]
Any news here?

**Reply by ematipico** [reply]
Please read the threads: https://github.com/biomejs/biome/discussions/1928#discussioncomment-10582258

**Comment by ice-cap0** [comment]
yeah this is tablestakes today imo.
going back to ESLint for now.

**Reply by schang124** [reply]
I am going back, too. *sigh*

**Comment by fireairforce** [comment]
So the remaining question is that  quick fix of `useExhaustiveDependencies` doesn't work well?

**Comment by PerryFinn** [comment]
Yes, I just encountered this issue while trying to migrate `eslint` and `prettier` to `biome`. I rely heavily on the automatic fix for `useEffect` dependencies, which saves a lot of valuable time for more important tasks.

**Comment by PerryFinn** [comment]
By the way, I'd like to ask, will the value generated by `useRef` also be forcibly specified as a dependency? This does not give me such a prompt when using `eslint`.

<img width="922" alt="image" src="https://github.com/user-attachments/assets/ba9da940-f495-48d9-8792-4d603f58f2aa">

Although it is possible to [configure](https://biomejs.dev/linter/rules/use-exhaustive-dependencies/#stable-results), I feel that having this rule by default might be more beginner-friendly.

**Comment by schang124** [comment]
Wow, I was excited to integrate biome, but figured it there is not quick fix for the react hook dependencies which is very convenient and also critical.
I am going back to eslint. No way I can use biome for my team unless this fearture is supported.

**Reply by kelani34** [reply]
Same, lol

**Reply by bufgix** [reply]
Same :(

**Comment by oddball** [comment]
@arendjr may I ask if any progress have been made on question number 1 ( of the original question ) ? Dying to get rid of eslint, but for me, biome is not useable until number 1 is automatic

**Comment by arendjr** [comment]
@oddball No, but I’ve created a task for this issue, so you can follow progress there: https://github.com/biomejs/biome/issues/5309

**Comment by oddball** [comment]
<img width="651" height="165" alt="Screenshot 2025-08-11 at 14 18 53" src="https://github.com/user-attachments/assets/3308d95b-2acc-48ea-86c9-d5477d136d0e" />


I gave biome another try. I am using version 2.1.4. Added a line to biome.json 
```json
{
  "$schema": "https://biomejs.dev/schemas/2.1.4/schema.json",
  "javascript": {
    "jsxRuntime": "reactClassic"
  },
  "linter": {
    "enabled": true,
    "domains": {
      "react": "recommended"
    },
    "rules": {
    ....
````

And using:



<img width="860" height="179" alt="Screenshot 2025-08-11 at 14 23 45" src="https://github.com/user-attachments/assets/24506f2d-7362-4e8e-b4e9-4894e5128ad2" />


@arendjr Is the quicFix not added to the vscode plugin? Cause vscode does not list an action to just fix an empty list

**Reply by arendjr** [reply]
They should work now, regardless of IDE used. If they don’t please file a separate issue with a reproduction.

**Reply by nicksama88** [reply]
I'm wondering if it's not showing up in the quickfix menu because it's an "unsafe" fix? I went ahead and made up a bug ticket with a reproduction though: https://github.com/biomejs/biome-vscode/issues/748. Really, really looking forward to this being available lol

**Comment by summerisgood1** [comment]
```
"linter": {
		"rules": {
			
			"correctness": {
				"useExhaustiveDependencies": {
					"level": "info",
					"fix": "safe"
				},
				
			},}}
```
Solution

**Reply by dangus21** [reply]
is this documented anywhere is their docs?

**Reply by nicksama88** [reply]
https://biomejs.dev/reference/vscode/#fix-on-save
https://biomejs.dev/linter/#configure-the-code-fix

The fix for `useExhaustiveDependencies` is definitely an "unsafe" one, though, so I don't think I'd suggest marking it as safe for the auto-fixer. I'd just apply the fix within the editor whenever you see an issue:

<img width="272" height="148" alt="image" src="https://github.com/user-attachments/assets/c99da774-f7e3-4540-b0cc-223e855a389b" />


<img width="614" height="165" alt="image" src="https://github.com/user-attachments/assets/44792eaa-41f1-4f63-89db-6c32f3cdf0d0" />

**Comment by EdonaTekkli** [comment]
<img width="940" height="438" alt="image" src="https://github.com/user-attachments/assets/037bc564-7ecd-445f-8c83-c2ba1b5b2fb9" />
Is this how the rule is treating the setter from `useState`?

**Reply by nicksama88** [reply]
I am on biome v 2.2.2 and extension v 3.2.0. It looks like, for me, the setter is properly understood to have a stable reference and isn't being called out by the linter.

<img width="419" height="142" alt="image" src="https://github.com/user-attachments/assets/cdee90d6-780b-46c9-a0b6-5652f324564f" />

<img width="344" height="137" alt="image" src="https://github.com/user-attachments/assets/913b6b6c-3fbb-401a-9621-18e2c81bee78" />
